datalad_next/tree.py

```diff
@@ -0,0 +1,1172 @@
+                #  count double.
+                self.node_count[node_type] += 1
+
+            yield node  # yield what the generator yielded
+
+    return _wrapper
+
+
+def yield_with_last_item(generator):
+    """Takes a generator and yields for each item, the item itself and
+    whether it is the last item in the sequence.
+
+    Returns
+    -------
+    Tuple[bool, Any]
+        A tuple (is_last_item, item)
+    """
+    prev_val = next(generator, None)
+    if prev_val is not None:
+        for current_val in generator:
+            yield False, prev_val
+            prev_val = current_val
+        yield True, prev_val
+
+
+def path_depth(path: Path, root: Path):
+    """Calculate directory depth of a path relative to the given root.
+
+    Can also be a negative integer if the path is a parent of the
+    tree root.
+
+    Returns
+    -------
+    int
+        Number of levels of the given path *below* the root (positive
+        integer) or *above* the tree root (negative integer)
+    """
+    if is_path_relative_to(path, root):
+        return len(path.relative_to(root).parts)
+    elif is_path_relative_to(root, path):
+        return - len(root.relative_to(path).parts)
+    else:
+        raise ValueError("Could not calculate directory depth: "
+                         f"'{path}' is not relative to the tree root "
+                         f"'{root}' (or vice-versa)")
+
+
+def is_empty_dir(path: Path):
+    return path.is_dir() and not any(path.iterdir())
+
+
+@lru_cache
+def is_dataset(path: Path):
+    """Fast dataset detection.
+
+    Infer that a directory is a dataset if it is either:
+
+    - installed, or
+    - not installed, but has an installed superdatset.
+
+    Only consider datalad datasets, not plain git/git-annex repos. Datasets
+    used for aggregating metadatata from subdatasets are also counted as
+    datasets, although they do not have a dataset ID themselves.
+
+    Symlinks pointing to datasets are not resolved, so will always return
+    False for symlinks. This prevents potentially detecting duplicate datasets
+    if the symlink and its target are both included in the tree.
+
+    Results are cached because the check is somewhat expensive and may
+    be run multiple times on the same path.
+
+    TODO: is there a way to detect a datalad dataset if it is not installed
+     and it is not a subdataset?
+
+    Parameters
+    ----------
+    path: Path
+        Path to directory to be identified as dataset or non-dataset
+    """
+    try:
+        if path.is_symlink():
+            # ignore symlinks even if pointing to datasets, otherwise we may
+            # get duplicate counts of datasets
+            lgr.debug("Path is a symlink, will not check if it points to a "
+                      f"dataset: '{path}'")
+            return False
+
+        if (path / ".datalad" / "config").is_file() or \
+                (path / ".datalad" / "metadata").is_dir():
```

Comment:
We can simplify this to:

```suggestion
        if (path / ".datalad" / "config").is_file():
```

The metadata dir is/will be no longer used in future datasets. The original implementation still needs to be adjusted too. This change has no significant performance impact.

Reply:
Good to know. Fixed in 070953dba.